src/ThreeEditor/components/Sidebar/tree/SidebarTreeItem.tsx

```diff
@@ -0,0 +1,226 @@
+import { Editor } from '../../../js/Editor';
+import { ISimulationObject } from '../../../util/SimulationBase/SimulationObject';
+import { NodeModel, TreeMethods } from '@minoru/react-dnd-treeview/dist/types';
+import {
+	canChangeName,
+	getRemoveCommand,
+	isRemovable
+} from '../../../util/hooks/useKeyboardEditorControls';
+import { isOutput } from '../../../util/Scoring/ScoringOutput';
+import { AddQuantityCommand } from '../../../js/commands/AddQuantityCommand';
+import Box from '@mui/material/Box';
+import { useSignal, useSmartWatchEditorState } from '../../../util/hooks/signals';
+import { useCallback, useEffect, useRef, useState } from 'react';
+
+export interface TreeItem {
+	id: number;
+	parent: number;
+	droppable: boolean;
+	text: string;
+	data: {
+		object: Object3D | SimulationObject3D;
+	};
+}
+
+function isHidable(object: Object3D | ISimulationObject) {
+	if ('notHidable' in object) {
+		return !object.notHidable;
+	}
+	return true;
+}
+
+export function SidebarTreeItem(props: {
+	treeRef: TreeMethods | null;
+	objectRefs: React.MutableRefObject<Map<string, HTMLDivElement>>;
+	node: NodeModel<{
+		object: Object3D | SimulationObject3D;
+	}>;
+	depth: number;
+	isOpen: boolean;
+	onToggle: () => void;
+	editor: Editor;
+}) {
+	const { objectRefs, node, depth, isOpen, onToggle, editor, treeRef } = props;
+	const inputRef = useRef<HTMLInputElement>(null);
+
+	const theme = useTheme();
+
+	const object = node.data!.object;
+
+	const { state: watchedObject } = useSmartWatchEditorState(editor, object);
+
+	const [mode, setMode] = useState<'view' | 'edit'>('view');
+
+	const popupState = usePopupState({ variant: 'popover', popupId: 'sidebarTreeItemMenu' });
+
+	const contextOptions = (() => {
+		const options: JSX.Element[] = [];
+
+		if (canChangeName(object)) {
+			options.push(
+				<MenuItem
+					onClick={() => {
+						popupState.close();
+						setMode('edit');
+					}}>
+					Rename
+				</MenuItem>
+			);
+		}
+
+		if (isOutput(object))
+			options.push(
+				<MenuItem
+					onClick={() => {
+						editor.execute(new AddQuantityCommand(editor, object));
+						if (!isOpen) onToggle();
+						popupState.close();
+					}}>
+					Add Quantity
+				</MenuItem>
+			);
+
+		if (isRemovable(object))
+			options.push(
+				<MenuItem
+					onClick={() => {
+						editor.execute(getRemoveCommand(editor, object));
+						popupState.close();
+					}}>
+					Delete
+				</MenuItem>
+			);
+
+		return options;
+	})();
+
+	useEffect(() => {
+		if (mode === 'edit') {
+			inputRef?.current?.focus();
+		}
+	}, [mode]);
+
+	const onObjectAdded = useCallback(
+		(newObject: Object3D) => {
+			if (newObject.parent === object) {
+				treeRef?.open(node.id);
+			}
+		},
+		[node.id, object, treeRef]
+	);
+
+	useSignal(editor, 'objectAdded', onObjectAdded);
+
+	return (
+		<>
+			<Box
+				ref={ref => objectRefs.current.set(object.uuid, ref! as HTMLDivElement)}
+				sx={{
+					'marginLeft': theme.spacing(depth * 2.5),
```

Comment:
@grzanka do we want tab to take space equal to the amount taken by extend/collapse button?
![image](https://user-images.githubusercontent.com/26445545/228286931-6dd3fbd0-a31a-4339-8844-ac948635b21e.png)
```suggestion
					'marginLeft': theme.spacing(depth * ???),
```

Reply:
That is a good idea. 4 * depth looks nice